src/scale_impls/encode.rs

```diff
@@ -13,793 +13,533 @@
-		None => {
-			return Err(EncodeError::VariantNotFound { actual: variant.clone(), expected: type_id })
-		}
-		Some(v) => v,
-	};
-
-	variant_type.index().encode_to(bytes);
-	encode_composite_fields(&variant.values, variant_type.fields(), type_id, types, bytes)
+    match ty.type_def() {
+        TypeDef::Tuple(tuple) if tuple.fields().len() == 1 => {
+            find_single_entry_with_same_repr(tuple.fields()[0].id(), types)
+        }
+        TypeDef::Composite(composite) if composite.fields().len() == 1 => {
+            find_single_entry_with_same_repr(composite.fields()[0].ty().id(), types)
+        }
+        _ => type_id,
+    }
 }
 
-fn encode_composite_fields<T: Clone>(
-	composite: &Composite<T>,
-	fields: &[Field<PortableForm>],
-	type_id: TypeId,
-	types: &PortableRegistry,
-	bytes: &mut Vec<u8>,
-) -> Result<(), EncodeError<T>> {
-	if fields.len() != composite.len() {
-		return Err(EncodeError::CompositeIsWrongLength {
-			actual: composite.clone(),
-			expected: type_id,
-			expected_len: fields.len(),
-		});
-	}
-
-	// 0 length? Nothing more to do!
-	if composite.is_empty() {
-		return Ok(());
-	}
-
-	// Does the type we're encoding to have named fields or not?
-	let is_named = fields[0].name().is_some();
-
-	match (composite, is_named) {
-		// If we provide named fields, and named fields are present on the target
-		// type, then we encode according to the names.
-		(Composite::Named(values), true) => {
-			// Match up named values with those of the type we're encoding to.
-			for field in fields.iter() {
-				let field_name = field.name().expect("field should be named; checked above");
-				let value = values.iter().find(|(n, _)| field_name == n).map(|(_, value)| value);
-
-				match value {
-					Some(value) => {
-						encode_value_as_type(value, field.ty(), types, bytes)?;
-					}
-					None => {
-						return Err(EncodeError::CompositeFieldIsMissing {
-							expected: type_id,
-							missing_field_name: field_name.clone(),
-						})
-					}
-				}
-			}
-			Ok(())
-		}
-		// If we provide named fields, and the target is unnamed fields, we'll just
-		// try to line them up in the order that they were given.
-		(Composite::Named(values), false) => {
-			for (field, (_name, value)) in fields.iter().zip(values) {
-				encode_value_as_type(value, field.ty(), types, bytes)?;
-			}
-			Ok(())
-		}
-		// If we provide unnamed fields, we don't care whether the target fields are
-		// named or not; we'll just line them up.
-		(Composite::Unnamed(values), _) => {
-			for (field, value) in fields.iter().zip(values) {
-				encode_value_as_type(value, field.ty(), types, bytes)?;
-			}
-			Ok(())
-		}
-	}
+// dig into a composite type and find a composite that is a candidate
+// for being encoded into a sequence.
+fn find_sequence_candidate<T>(value: &'_ Composite<T>) -> &'_ Composite<T> {
+    if value.len() != 1 {
+        return value;
+    }
+
+    let inner_value = value.values().next().unwrap();
```

Comment:
same here, I would prefer either handle the iterator properly or use expect

Reply:
Expect used :)

(elsewhere I did something like `if let (Some(val), None) = (values.next(), values.next())..` but this version feels a little clearer despite the expect (but I could easily go either way!))

Reply:
Actually I changed my mind and went with the match version to avoid expects, and in this case it cuts dow nthe code anyway :)